fdbserver/ClusterController.actor.cpp

```diff
@@ -1691,6 +1700,196 @@ class ClusterControllerData {
+					continue;
+				}
+				degradationCounts[degradedServer]++;
+			}
+		}
+
+		std::unordered_set<NetworkAddress> excludeServers;
+		for (const auto& [server, degradationCount] : degradationCounts) {
+			if (degradationCount <= SERVER_KNOBS->MAX_PEER_DEGRADATION_TO_EXCLUDE) {
+				excludeServers.insert(server);
+				continue;
+			}
+		}
+		return excludeServers;
+	}
+
+	void updatePeerHealth(const UpdatePeerHealthRequest& req) {
+		double currentTime = now();
+		if (req.degradedPeers.empty()) {
+			return;
+		}
+
+		auto it = degradedPeers.find(req.selfAddress);
+		if (it == degradedPeers.end()) {
+			degradedPeers[req.selfAddress] = {};
+			for (const auto& peerAddress : req.degradedPeers) {
+				degradedPeers[req.selfAddress][peerAddress] = currentTime;
+			}
+		} else {
+			std::unordered_set<NetworkAddress> recoveredPeers;
+			for (const auto& [peer, lastUpdateTime] : degradedPeers[req.selfAddress]) {
+				recoveredPeers.insert(peer);
+			}
+			for (const auto& peer : req.degradedPeers) {
+				if (recoveredPeers.find(peer) != recoveredPeers.end()) {
+					recoveredPeers.erase(peer);
+				}
+			}
+			// Now recoveredPeers stores recovered peers;
+			for (const auto& peer : recoveredPeers) {
+				degradedPeers[req.selfAddress].erase(peer);
+			}
+
+			for (const auto& peer : req.degradedPeers) {
+				if (degradedPeers[req.selfAddress].find(peer) == degradedPeers[req.selfAddress].end()) {
+					degradedPeers[req.selfAddress][peer] = currentTime;
+				}
+			}
+		}
+	}
+
+	bool TriggerRecoveryCausedByDegradedServers() {
+		const ServerDBInfo dbi = db.serverInfo->get();
+
+		if (dbi.recoveryState < RecoveryState::ACCEPTING_COMMITS) {
+			return false;
+		}
+
+		// Do not trigger better master exists if the cluster controller is excluded, since the master will change
+		// anyways once the cluster controller is moved
+		if (id_worker[clusterControllerProcessId].priorityInfo.isExcluded) {
+			return false;
+		}
+
+		if (db.config.regions.size() > 1 && db.config.regions[0].priority > db.config.regions[1].priority &&
+		    db.config.regions[0].dcId != clusterControllerDcId.get() && versionDifferenceUpdated &&
+		    datacenterVersionDifference < SERVER_KNOBS->MAX_VERSION_DIFFERENCE) {
+			checkRegions(db.config.regions);
+		}
+
+		// Get master process
+		auto masterWorker = id_worker.find(dbi.master.locality.processId());
+		if (masterWorker == id_worker.end()) {
+			return false;
+		}
+
+		// Get tlog processes
+		std::vector<WorkerDetails> tlogs;
+		std::vector<WorkerDetails> remote_tlogs;
+		std::vector<WorkerDetails> satellite_tlogs;
+		std::vector<WorkerDetails> log_routers;
+		std::set<NetworkAddress> logRouterAddresses;
+		std::vector<WorkerDetails> backup_workers;
+		std::set<NetworkAddress> backup_addresses;
+
+		for (auto& logSet : dbi.logSystemConfig.tLogs) {
+			for (auto& it : logSet.tLogs) {
+				auto tlogWorker = id_worker.find(it.interf().filteredLocality.processId());
+				if (tlogWorker == id_worker.end())
+					return false;
```

Comment:
Can you elaborate  on why return false here? 

nitt: we use {} for a single line body. so can you add {} around `return false`?